### PR TITLE
Add Batch Audit Template Import

### PR DESCRIPTION
**Pull request by perryr16**
#### Any background context you want to provide?
SEED currently has functionality to update an individual property with [Audit Template](https://buildingenergyscore.energy.gov/rp/buildings) building data. The function leverages the Audit Template API endpoint [GET /api/v2/building_sync/download/rp/buildings/:id](https://api.labworks.org/apidoc/v2/building_sync/download.html). The returned XML is used to create a new PropertyState and associate it with the current PropertyView.

#### What's this PR do?
Adds the ability to batch update properties with data from audit template.

From the Data page, a user may now import Audit Template Buildings via the Add Data Files button. A request is sent to the linked Audit Template account to get all buildings associated with the account details at [GET /api/v2/rp/buildings](https://api.labworks.org/apidoc/v2/buildings/index.html). Backend logic determines which audit tempatae buildings match existing properties for the selected cycle. Matching is solely based on the PropertyState field `audit_template_building_id`. The results are displayed to the user where they may sort/filter results and select which properties to continue to update. Selected properties are looped through and XML files are fetched from the Audit Template API endpoint `GET /api/v2/building_sync/download/rp/buildings/:id`. PropertyViews are updated with the corresponding XMLs.

All Audit Template requests are made within the `seed_celery` container. 

The current implementation can lead to very long wait times. For example, a request to get a building xml for a single building takes ~ 5 seconds. Updating 1000 properties would take about an hour and a half to complete.

The only property that exists in both the PropertyState and the Audit Template Tool is `audit_template_building_id` (`id` on Audit Template). This requires SEED properties to be updated with correct `audit_tempalte_building_id`s which can be time consuming.

If the matching criteria were to change, the audit template get_buildings response must be updated to include properties like `pm_property_id`, `ubid`, or `custom_id`. Below is the current response structure.

```
GET /api/v2/rp/buildings
[{
    "id": 369,
    "name": "New Buildng",
    "owner": {
        "id": 877,  "email": "[test](mailto:test@nrel.gov)", "first_name": "test",
        "last_name": "test",  "organization": { "id": 9, 
        "name": "National Renewable Energy Laboratory - SEED Project" }, 
        "site_admin": false,  "type": "User" 
     },
    "report": {
      "id": 2, "name": "New York City Energy Efficiency Report",
      "city": { "id": 2, "name": "New York", "state": "NY" }
    },
   "status": "Editing",
   "submissions": [],
   "created_at": "2018-08-22T12:21:43.519-07:00",
   "updated_at": "2023-04-04T12:31:04.100-07:00",
   "html_url": "<html_url>",
   "submission_history_html_url": "<history_url>",
   "xml_url":  "<xml_url>"
  },
   ….
]
```
#### Known Issues 
Filtering the Audit template buildings only supports exact matches at the moment, not '<10', '>20'

#### How should this be manually tested?
Add audit_template_building_ids to existing properties with building ids from your Audit Template Tool account. 
Data > Add Data Files > Import Audit Template Buildings > Confirm > select properties > Update
You should see an informative progress bar and a quick summary of new property views added. Check the property detail of the updated properties to ensure a new building file has been uploaded (`at_{audit_tempalte_building_id}_{YYYYMMDD_Hms}`) and the properties attributes have been updated with Audit Template data

#### What are the relevant tickets?
#3770

#### Screenshots (if appropriate)

Screenshot 2023-07-26 at 9 41 18 AM
Screenshot 2023-07-26 at 9 41 23 AM
Screenshot 2023-07-26 at 9 41 34 AM

Screenshot 2023-07-26 at 9 43 03 AM
Screenshot 2023-07-26 at 9 43 13 AM
Screenshot 2023-07-26 at 9 43 24 AM

### CODE REVIEW

File: seed/views/v3/audit_template.py [hunk 1]
```diff
@@ -29,3 +32,109 @@ def get_building_xml(self, request, pk):
...
+
+        if progress_data is None:
+            return JsonResponse({
+                'success': False,
+                'message': 'Unexpected Error'
+            }, status=400)
+
+        return JsonResponse(progress_data)
+
+    def validate_properties(self, properties):
```
Comment (perryr16) [inline seed/views/v3/audit_template.py:85]: There's probably a better way to do this

File: seed/audit_template/audit_template.py [hunk 2]
```diff
@@ -64,3 +95,56 @@ def get_api_token(self):
...
+        response = requests.request("GET", url, headers=headers)
+        if response.status_code != 200:
+            return None, f'Expected 200 response from Audit Template get_buildings but got {response.status_code}: {response.content}'
+    except Exception as e:
+        return None, f'Unexpected error from Audit Template: {e}'
+    at_buildings = response.json()
+    result = []
+    for b in at_buildings:
+        # Only update properties that have been recently updated on Audit Template
+        at_updated = datetime.fromisoformat(b['updated_at']).strftime("%Y-%m-%d %I:%M %p")
```
Comment (perryr16) [inline seed/audit_template/audit_template.py:112]: Audit Template returns iso format dates for the `updated` attribute. SEED formats updated/created as `YYYY-MM-DD HH:mm:ss p`. I think we should keep date formats consistent across seed

File: seed/views/v3/properties.py [hunk 3]
```diff
@@ -1338,11 +1342,34 @@ def update_with_building_sync(self, request, pk):
...
+            blob = ContentFile(property['xml'], name=f'at_{property["audit_template_building_id"]}_{formatted_time}.xml')
+            response = self._update_with_building_sync(blob, 1, org_id, cycle_id, property['property_view'], property['updated_at'])
+            response = json.loads(response.content)
+            results['success' if response['success'] else 'faulure'] += 1
+
+            progress_data.step('Updating Properties...')
+
+        progress_data.finish_with_success(results)
+
+    def _update_with_building_sync(self, the_file, file_type, organization_id, cycle_id, view_id, at_updated=False):
```
Comment (perryr16) [inline seed/views/v3/properties.py:1370]: this function gets called when an update from the property detail is triggered, or when a batch_update from the data page is triggered. When the batch update is triggered a get_buildings request is first made. This returns `updated` from Audit Template, it then loops through selected buildings and `_update_with_building_sync`.

The property detail only triggers the `get_building_xml` endpoint which does not return the `updated` attribute from Audit Template. Getting the `updated` would require an additional call to Audit Template.

File: seed/views/v3/properties.py [hunk 4]
```diff
@@ -1376,6 +1403,17 @@ def update_with_building_sync(self, request, pk):
...
+            if at_updated:
+                new_pv_state.extra_data.update({'at_updated_at': at_updated})
+                new_pv_state.save()
+                Column.objects.get_or_create(
+                    is_extra_data=True,
+                    column_name='at_updated_at',
+                    display_name='Audit Template Updated',
+                    organization=cycle.organization,
+                    table_name='PropertyState',
+                )
```
Comment (perryr16) [inline seed/views/v3/properties.py:1415]: update extra data to have at_updated field and create a column if it DNE to allow the frontend to share the value

**Review by kflemin — COMMENTED** [review]
🏆

**Comment by perryr16** [comment]
This work is included in PR #4215

**Pull request closed without merging**